packages/shorebird_cli/lib/src/commands/run_command.dart

```diff
@@ -0,0 +1,166 @@
+      return ExitCode.noUser.code;
+    }
+
+    final shorebirdEnginePath = p.join(
+      Directory.current.path,
+      '.shorebird',
+      'engine',
+    );
+    final shorebirdEngine = Directory(shorebirdEnginePath);
+    final shorebirdEngineCache = File(
+      p.join(Directory.current.path, '.shorebird', 'cache', 'engine.zip'),
+    );
+
+    if (!shorebirdEngineCache.existsSync()) {
+      final downloadEngineProgress = _logger.progress(
+        'Downloading shorebird engine',
+      );
+      try {
+        final codePushApiClient = _buildCodePushApiClient(
+          apiKey: session.apiKey,
+        );
+        await _downloadShorebirdEngine(
+          codePushApiClient,
+          shorebirdEngineCache.path,
+        );
+        downloadEngineProgress.complete();
+      } catch (error) {
+        downloadEngineProgress.fail(
+          'Failed to download shorebird engine: $error',
+        );
+        return ExitCode.software.code;
+      }
+    }
+
+    if (!shorebirdEngine.existsSync()) {
+      final buildingEngine = _logger.progress(
+        'Building shorebird engine',
+      );
+      try {
+        await _extractShorebirdEngine(
+          shorebirdEngineCache.path,
+          shorebirdEngine.path,
+        );
+        buildingEngine.complete();
+      } catch (error) {
+        buildingEngine.fail(
+          'Failed to build shorebird engine: $error',
+        );
+        return ExitCode.software.code;
+      }
+    }
+
+    _logger.info('Running app...');
+    final process = await _startProcess(
+      'flutter',
+      [
+        'run',
+        // Eventually we should support running in both debug and release mode.
+        '--release',
+        '--local-engine-src-path',
+        shorebirdEnginePath,
+        '--local-engine',
+        // This is temporary because the Shorebird engine currently
+        // only supports Android arm64.
+        'android_release_arm64',
+        if (argResults?.rest != null) ...argResults!.rest
+      ],
+      runInShell: true,
+    );
+
+    process.stdout.listen((event) {
+      _logger.info(utf8.decode(event));
+    });
+    process.stderr.listen((event) {
+      _logger.err(utf8.decode(event));
+    });
+
+    return process.exitCode;
+  }
+}
+
+Future<void> _downloadShorebirdEngine(
+  ShorebirdCodePushApiClient codePushClient,
+  String path,
+) async {
+  final engine = await codePushClient.downloadEngine('latest');
+  final targetFile = File(path);
+
+  if (targetFile.existsSync()) targetFile.deleteSync(recursive: true);
+
+  targetFile.createSync(recursive: true);
+  await targetFile.writeAsBytes(engine, flush: true);
+}
+
+Future<void> _extractShorebirdEngine(
+  String archivePath,
+  String targetPath,
+) async {
+  final targetDirectory = Directory(targetPath);
+
+  if (targetDirectory.existsSync()) targetDirectory.deleteSync(recursive: true);
+
+  targetDirectory.createSync(recursive: true);
+
+  await compute(
```

Comment:
Could this function end up getting called twice and causing racing computes?  I wonder if we shouldn't just do this sync?

Reply:
If we do it sync then it'll block and the progress reporting will freeze up (and right now it takes me ~20s) which is long enough to where it might look like we locked up your shell 😛 

Lmk what you think.